### PR TITLE
[Merged by Bors] - feat: change definition of `HasFTaylorSeries` to take a parameter in `WithTop ℕ∞` instead of `ℕ∞`

### PR DESCRIPTION
**Pull request by sgouezel**
For now this is useless (and even counter-productive, as it means we have to add some casts in some places), but it will make it possible to integrate analytic functions in the smooth hierarchy in #17152.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/Calculus/ContDiff/Basic.lean [hunk 1]
```diff
@@ -220,16 +220,16 @@ theorem ContDiff.continuousLinearMap_comp {f : E → F} (g : F →L[𝕜] G) (hf
 /-- The iterated derivative within a set of the composition with a linear map on the left is
 obtained by applying the linear map to the iterated derivative. -/
 theorem ContinuousLinearMap.iteratedFDerivWithin_comp_left {f : E → F} (g : F →L[𝕜] G)
-    (hf : ContDiffOn 𝕜 n f s) (hs : UniqueDiffOn 𝕜 s) (hx : x ∈ s) {i : ℕ} (hi : (i : ℕ∞) ≤ n) :
+    (hf : ContDiffOn 𝕜 n f s) (hs : UniqueDiffOn 𝕜 s) (hx : x ∈ s) {i : ℕ} (hi : i ≤ n) :
     iteratedFDerivWithin 𝕜 i (g ∘ f) s x =
       g.compContinuousMultilinearMap (iteratedFDerivWithin 𝕜 i f s x) :=
   (((hf.ftaylorSeriesWithin hs).continuousLinearMap_comp g).eq_iteratedFDerivWithin_of_uniqueDiffOn
-    hi hs hx).symm
+    (by exact_mod_cast hi) hs hx).symm
```
Comment (Command-Master) [inline Mathlib/Analysis/Calculus/ContDiff/Basic.lean:227]:
```suggestion
    (mod_cast hi) hs hx).symm
```

File: Mathlib/Analysis/Calculus/ContDiff/Basic.lean [hunk 2]
```diff
@@ -371,11 +371,11 @@ theorem ContDiff.comp_continuousLinearMap {f : E → F} {g : G →L[𝕜] E} (hf
 obtained by composing the iterated derivative with the linear map. -/
 theorem ContinuousLinearMap.iteratedFDerivWithin_comp_right {f : E → F} (g : G →L[𝕜] E)
     (hf : ContDiffOn 𝕜 n f s) (hs : UniqueDiffOn 𝕜 s) (h's : UniqueDiffOn 𝕜 (g ⁻¹' s)) {x : G}
-    (hx : g x ∈ s) {i : ℕ} (hi : (i : ℕ∞) ≤ n) :
+    (hx : g x ∈ s) {i : ℕ} (hi : i ≤ n) :
     iteratedFDerivWithin 𝕜 i (f ∘ g) (g ⁻¹' s) x =
       (iteratedFDerivWithin 𝕜 i f s (g x)).compContinuousLinearMap fun _ => g :=
   (((hf.ftaylorSeriesWithin hs).compContinuousLinearMap g).eq_iteratedFDerivWithin_of_uniqueDiffOn
-    hi h's hx).symm
+    (by exact_mod_cast hi) h's hx).symm
```
Comment (Command-Master) [inline Mathlib/Analysis/Calculus/ContDiff/Basic.lean:378]:
```suggestion
    (mod_cast hi) h's hx).symm
```

File: Mathlib/Analysis/Calculus/ContDiff/Basic.lean (hunk 1 above)
Comment (sgouezel) [reply Mathlib/Analysis/Calculus/ContDiff/Basic.lean]: Nice, thanks! I have adopted this throughout the whole PR.

File: Mathlib/Analysis/Calculus/ContDiff/RCLike.lean [hunk 3]
```diff
@@ -24,12 +24,13 @@ section Real
   its extension fields such as `ℂ`).
 -/
 
-variable {n : ℕ∞} {𝕂 : Type*} [RCLike 𝕂] {E' : Type*} [NormedAddCommGroup E'] [NormedSpace 𝕂 E']
-  {F' : Type*} [NormedAddCommGroup F'] [NormedSpace 𝕂 F']
+variable {n : ℕ∞} {𝕂 : Type*} [RCLike 𝕂] {E' : Type*} [NormedAddCommGroup E']
+  [NormedSpace 𝕂 E'] {F' : Type*} [NormedAddCommGroup F'] [NormedSpace 𝕂 F']
```
Comment (Command-Master) [inline Mathlib/Analysis/Calculus/ContDiff/RCLike.lean:28]: What's the reason for this change?

File: Mathlib/Data/ENat/Basic.lean [hunk 4]
```diff
@@ -283,13 +294,12 @@ lemma one_le_iff_ne_zero_withTop {n : WithTop ℕ∞} :
...
 
-lemma add_one_pos : 0 < n + 1 :=
-  succ_def n ▸ Order.bot_lt_succ n
+lemma nat_le_of_infty_le_withTop {N : WithTop ℕ∞} (hN : (⊤ : ℕ∞) ≤ N) (n : ℕ) : n ≤ N :=
+  le_trans (by exact_mod_cast le_top) hN
 
-lemma add_lt_add_iff_right {k : ℕ∞} (h : k ≠ ⊤) : n + k < m + k ↔ n < m :=
-  WithTop.add_lt_add_iff_right h
+lemma nat_lt_of_infty_le_withTop {N : WithTop ℕ∞} (hN : (⊤ : ℕ∞) ≤ N) (n : ℕ) : n < N :=
+  lt_of_lt_of_le (by exact_mod_cast lt_add_one n) (nat_le_of_infty_le_withTop hN (n + 1))
```
Comment (Command-Master) [inline Mathlib/Data/ENat/Basic.lean:301]: Does `mod_cast` work here?

File: Mathlib/Data/ENat/Basic.lean [hunk 5]
```diff
@@ -261,6 +261,17 @@ theorem nat_induction {P : ℕ∞ → Prop} (a : ℕ∞) (h0 : P 0) (hsuc : ∀
...
+lemma add_one_pos : 0 < n + 1 :=
+  succ_def n ▸ Order.bot_lt_succ n
+
+lemma add_lt_add_iff_right {k : ℕ∞} (h : k ≠ ⊤) : n + k < m + k ↔ n < m :=
+  WithTop.add_lt_add_iff_right h
+
+lemma add_lt_add_iff_left {k : ℕ∞} (h : k ≠ ⊤) : k + n < k + m ↔ n < m :=
+  WithTop.add_lt_add_iff_left h
+
+section withTop_enat
```
Comment (Command-Master) [inline Mathlib/Data/ENat/Basic.lean:273]: Section names aren't all that important, but I think simply `WithTop` would be a better namespace here, as it's already in the `ENat` namespace

File: Mathlib/Data/ENat/Basic.lean (hunk 4 above)
Comment (sgouezel) [reply Mathlib/Data/ENat/Basic.lean:301]: Yes it does, thanks!

File: Mathlib/Data/ENat/Basic.lean (hunk 5 above)
Comment (sgouezel) [reply Mathlib/Data/ENat/Basic.lean:273]: I'd rather keep `withTop_enat`: otherwise, the section could be interpreted as a discussion on the relationship between `ENat` and `WithTop Nat`, and ways to go back and forth.

File: Mathlib/Analysis/Calculus/ContDiff/RCLike.lean (hunk 3 above)
Comment (sgouezel) [reply Mathlib/Analysis/Calculus/ContDiff/RCLike.lean:28]: No reason. Reverted.

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge
